File: app/src/main/java/com/popalay/cardme/ui/base/widgets/CharacterWrapTextView.java
package com.popalay.cardme.ui.base.widgets;

import android.content.Context;
import android.graphics.Rect;
import android.support.v7.widget.AppCompatTextView;
import android.util.AttributeSet;
import android.util.Log;

public class CharacterWrapTextView extends AppCompatTextView {

    private static final String TAG = "CharacterWrapTextView";

    public CharacterWrapTextView(Context context) {
        super(context);
    }

    public CharacterWrapTextView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public CharacterWrapTextView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
    }

    @Override
    public void setText(CharSequence text, BufferType type) {
        applyLetterSpacing(text);
        super.setText(text, type);
    }

    private void applyLetterSpacing(CharSequence text) {
        final Rect bounds = new Rect();
        getPaint().getTextBounds(text.toString(), 0, text.length(), bounds);
        final int textWidth = bounds.width();
        if (textWidth == 0) {
            return;
        }
        post(() -> {
            final int width = getWidth();
            final float spacing = 1f - (float) textWidth / width - 0.1f;
            Log.d(TAG, "applyLetterSpacing: " + text.toString() + " " + width + " " + textWidth + " " + spacing);
            setLetterSpacing(spacing);
        });

    }

}

File: presentation/src/main/kotlin/com/popalay/cardme/presentation/widget/CharacterWrapTextView.kt
package com.popalay.cardme.presentation.widget

import android.content.Context
import android.graphics.Paint
import android.support.v7.widget.AppCompatTextView
import android.util.AttributeSet
import android.view.View

class CharacterWrapTextView(context: Context,
                            attrs: AttributeSet? = null,
                            defStyleAttr: Int = 0
) : AppCompatTextView(context, attrs, defStyleAttr) {

    constructor(context: Context, attrs: AttributeSet) : this(context, attrs, 0)

    private val testPaint: Paint by lazy { Paint(paint) }

    private fun refitText(text: String, textWidth: Int) {
        if (textWidth <= 0) return

        val targetWidth = textWidth - paddingLeft - paddingRight
        var hi = 1f
        var lo = 0f

        while (hi - lo > 0.1f) {
            val size = (hi + lo) / 2
            testPaint.letterSpacing = size
            val textWidthCalculated = testPaint.measureText(text, 0, text.length)
            if (textWidthCalculated >= targetWidth) hi = size else lo = size
        }

        if (lo == letterSpacing) return

        letterSpacing = lo
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        val parentWidth = View.MeasureSpec.getSize(widthMeasureSpec)
        val height = measuredHeight
        refitText(text.toString(), parentWidth)
        this.setMeasuredDimension(parentWidth, height)
    }

    override fun onTextChanged(text: CharSequence, start: Int, before: Int, after: Int) {
        refitText(text.toString(), width)
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        if (w != oldw) refitText(text.toString(), w)
    }
}